### PR TITLE
snowflake: schema evolution

### PR DESCRIPTION
**Pull request by rockwotj**
Support schema evolution for Redpanda Connect's new Snowflake output.

### CODE REVIEW

File: internal/impl/snowflake/streaming/schema_errors.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+package streaming
+
+import "fmt"
+
+// SchemaMismatchError occurs when the user provided data has data that
+// doesn't match the schema *and* the table can be evolved to accomidate
```
Comment (mihaitodor) [inline internal/impl/snowflake/streaming/schema_errors.go:16]: Typo: `accomidate`

File: internal/impl/snowflake/streaming/compat.go [hunk 2]
```diff
@@ -150,6 +150,27 @@ func normalizeColumnName(name string) string {
...
 }
 
+// quoteColumnName escapes an object identifier according to the
+// rules in Snowflake.
+//
+// https://docs.snowflake.com/en/sql-reference/identifiers-syntax
+func quoteColumnName(name string) string {
+	var quoted strings.Builder
+	// Default to assume we're just going to add quotes and there won't
+	// be any double quotes inside the string that need escaped.
```
Comment (mihaitodor) [inline internal/impl/snowflake/streaming/compat.go:160]: Typo: needs

File: internal/impl/snowflake/output_snowflake_streaming.go [hunk 3]
```diff
@@ -456,6 +541,129 @@ func (o *snowflakeStreamerOutput) WriteBatch(ctx context.Context, batch service.
...
+		"name":  col.RawName(),
+		"value": col.Value(),
+	})
+	out, err := msg.BloblangQuery(o.schemaEvolutionMapping)
+	if err != nil {
+		return fmt.Errorf("unable to compute new column type for %s: %w", col.ColumnName(), err)
+	}
+	v, err := out.AsBytes()
+	if err != nil {
+		return fmt.Errorf("unable to compute new column type for %s: %w", col.ColumnName(), err)
```
Comment (mihaitodor) [inline internal/impl/snowflake/output_snowflake_streaming.go:566]: Can't distinguish between these errors

File: internal/impl/snowflake/output_snowflake_streaming.go [hunk 4]
```diff
@@ -456,6 +541,129 @@ func (o *snowflakeStreamerOutput) WriteBatch(ctx context.Context, batch service.
...
+		// that need polling to wait until they finish (results are made async when execution is longer
+		// than 45 seconds).
+		Timeout:  30,
+		Database: o.db,
+		Schema:   o.schema,
+		Role:     o.role,
+		Bindings: map[string]streaming.BindingValue{
+			"1": {Type: "TEXT", Value: o.table},
+		},
+	})
```
Comment (mihaitodor) [inline internal/impl/snowflake/output_snowflake_streaming.go:593]: Could we have a wrapper func so we don't duplicate this code & comments in `MigrateNotNullColumn()`?

**Review by mihaitodor — APPROVED** [review]
Thank you @rockwotj! LGTM, just left a few small nitpicks. Feel free to 🐑 🚀

File: internal/impl/snowflake/streaming/schema_errors.go (hunk 1 above)
Comment (rockwotj) [reply internal/impl/snowflake/streaming/schema_errors.go:16]: 🙏

File: internal/impl/snowflake/output_snowflake_streaming.go (hunk 3 above)
Comment (rockwotj) [reply internal/impl/snowflake/output_snowflake_streaming.go:566]: Good call.

File: internal/impl/snowflake/streaming/compat.go (hunk 2 above)
Comment (rockwotj) [reply internal/impl/snowflake/streaming/compat.go:160]: Thanks.

File: internal/impl/snowflake/output_snowflake_streaming.go (hunk 4 above)
Comment (rockwotj) [reply internal/impl/snowflake/output_snowflake_streaming.go]: Yes - thanks it's much easier to read now :)

**Pull request merged**